theano/gof/destroyhandler.py

```diff
@@ -945,6 +945,20 @@ def orderings(self, fgraph):
                 raise InconsistencyError("Attempting to destroy indestructible variables: %s" %
                         illegal_destroy)
 
+            # We want to compute a list of nodes that may never be computed,
+            # so that we do not force their computations (they may be uncomputeable).
+            # For that, we start by a list of lazy nodes (ifelse), and use io_toposort
+            # to find out which variables will always be computed, even if the lazy
+            # ones are not.
+            lazy_outputs = [var for var in fgraph.variables
+                            if var.owner and isinstance(var.owner, theano.ifelse.IfElse)]
+            always_computed = graph.io_toposort(fgraph.outputs, lazy_outputs)
```

Comment:
My idea was to not disable inplace Ops between the output of lazy operations and the output of the graph.
If all outputs are ifelse nodes, then it would disable it completely.
I know we have to be more clever than that, in particular:
- the "condition" input of ifelse is always computed if the ifelse is computed,
- if a chain of node is only used in one "branch" of an ifelse, then they can work inplace on one another,
- if a node is used in both branches of the same ifelse, then it can work inplace on an earlier one

The first and last points are specific to ifelse, and not other potential lazy ops, so that logic should probably be expressed in the IfElse Op instead, but I don't know how to specify all that.